### PR TITLE
Impelement OpenCL enum generation

### PR DESCRIPTION
**Pull request by Alexx999**
This should add some basic enum generation support for OpenCL aside from the umbrella CLEnum type.

The initial motivation is that some of the enum types have to be ulong-based causing OpenCL API calls to fail in funny ways.

### CODE REVIEW

File: src/OpenCL/Silk.NET.OpenCL/Enums/CLEnum.gen.cs
Comment (Perksey) [inline src/OpenCL/Silk.NET.OpenCL/Enums/CLEnum.gen.cs:860]: Double space?

File: src/OpenCL/Silk.NET.OpenCL/Enums/IntelAdvancedMotionEstimationIntraSearchPredictionModesBufferChromaBlock.gen.cs
Comment (Perksey) [inline src/OpenCL/Silk.NET.OpenCL/Enums/IntelAdvancedMotionEstimationIntraSearchPredictionModesBufferChromaBlock.gen.cs:13]: lmao this is hilarious

File: src/OpenCL/Silk.NET.OpenCL/Enums/LoaderLayers.gen.cs
Comment (Perksey) [inline src/OpenCL/Silk.NET.OpenCL/Enums/LoaderLayers.gen.cs:13]: This is a breaking change. While we're happy to break if the upstream specification dictates it, we generally always have ungrouped enums available thus if it's possible I'd prefer to either:
- have the grouped enum in addition to the ungrouped vendor-wide enum
- stop this change entirely

If 1 is not too much trouble ;)

File: src/OpenCL/Silk.NET.OpenCL/Enums/SemaphoreTypeKhr.gen.cs
Comment (Perksey) [inline src/OpenCL/Silk.NET.OpenCL/Enums/SemaphoreTypeKhr.gen.cs:13]: Another one here too.

**Review by Perksey — CHANGES_REQUESTED** [review]

**Comment by Perksey** [comment]
Thank you so much for this contribution! I've already commended your efforts plenty in Discord, but we are genuinely delighted to see community efforts being directed here! We cannot thank you enough.

I suppose one final comment, in addition to the nitpicks above, is: is there any way to make the GroupOverloader light up so that these enums are properly integrated to the rest of the library?

File: src/Core/Silk.NET.BuildTools/Converters/Readers/OpenCLReader.cs
Comment (Alexx999) [reply src/Core/Silk.NET.BuildTools/Converters/Readers/OpenCLReader.cs:1003]: Basically, "(Grouped)" are separate enums. as opposed to the large root CLEnum

File: src/Core/Silk.NET.BuildTools/Converters/Readers/OpenCLReader.cs
Comment (Alexx999) [reply src/Core/Silk.NET.BuildTools/Converters/Readers/OpenCLReader.cs:1058]: I've submitted a PR to Khronos, so this wouldn't have to stay for long anyway

Comment (Alexx999) [reply src/Core/Silk.NET.BuildTools/Converters/Readers/OpenCLReader.cs]: OK, I see what you're saying now, switched to the JSON-based renamer

File: src/OpenCL/Silk.NET.OpenCL/Enums/LoaderLayers.gen.cs
Comment (Alexx999) [reply src/OpenCL/Silk.NET.OpenCL/Enums/LoaderLayers.gen.cs]: Yeah, found the issue, those empty enums should be back in business now

File: src/OpenCL/Silk.NET.OpenCL/Enums/CLEnum.gen.cs
Comment (Alexx999) [reply src/OpenCL/Silk.NET.OpenCL/Enums/CLEnum.gen.cs]: Good catch

**Comment by Alexx999** [comment]
Actually I have 2 more PRs pretty much ready - these changes don't actually achieve my initial goal (namely, incorrect stack size for some calls due to incorrectly sized stuff), I just don't want to dump too much code at once
Next up, constants handling and finally a typemapper change that builds on top of these 2 to achieve correct codegen (or, at least, "more correct").

File: src/Core/Silk.NET.BuildTools/Converters/Readers/OpenCLReader.cs
Comment (Perksey) [reply src/Core/Silk.NET.BuildTools/Converters/Readers/OpenCLReader.cs:999]: Gotcha, thanks.

File: src/Core/Silk.NET.BuildTools/Converters/Readers/OpenCLReader.cs
Comment (Perksey) [inline src/Core/Silk.NET.BuildTools/Converters/Readers/OpenCLReader.cs:955]: Can we use this data to add a `[Flags]` attribute to the relevant enums? Otherwise you get compiler warnings when doing bitwise stuff.

**Review by Perksey — CHANGES_REQUESTED** [review]

File: src/Core/Silk.NET.BuildTools/Converters/Readers/OpenCLReader.cs
Comment (Alexx999) [reply src/Core/Silk.NET.BuildTools/Converters/Readers/OpenCLReader.cs:955]: Yeah, I was also thinking about this.
I wonder what the process should be for larger PRs like this: keep adding stuff until it's all good?

Comment (Perksey) [reply src/Core/Silk.NET.BuildTools/Converters/Readers/OpenCLReader.cs:955]: Generally we do prefer smaller PRs, but in the world of BuildTools this is still only a small PR so your call!

File: src/Core/Silk.NET.BuildTools/Converters/Readers/OpenCLReader.cs
Comment (Perksey) [inline src/Core/Silk.NET.BuildTools/Converters/Readers/OpenCLReader.cs:808]: Can you add a comment explaining why these can't make the first pass?

File: src/Core/Silk.NET.BuildTools/Converters/Readers/OpenCLReader.cs
Comment (Perksey) [inline src/Core/Silk.NET.BuildTools/Converters/Readers/OpenCLReader.cs:863]: Can this not be hardcoded

File: src/Core/Silk.NET.BuildTools/Converters/Readers/OpenCLReader.cs
Comment (Perksey) [inline src/Core/Silk.NET.BuildTools/Converters/Readers/OpenCLReader.cs:861]: Can these not be hardcoded

File: src/Core/Silk.NET.BuildTools/Converters/Readers/OpenCLReader.cs
Comment (Perksey) [inline src/Core/Silk.NET.BuildTools/Converters/Readers/OpenCLReader.cs:873]: All in all I'm extremely uncomfortable using the "comment" attribute meaningfully. If there is some data in the comment attributes that we need, I'd recommend this be augmented & extracted to its own attribute in the upstream specs. Or at least this has been the approach with OpenGL.

File: src/Core/Silk.NET.BuildTools/Converters/Readers/OpenCLReader.cs
Comment (Perksey) [inline src/Core/Silk.NET.BuildTools/Converters/Readers/OpenCLReader.cs:1329]: Can this be made a record struct? This is very verbose.

File: src/Core/Silk.NET.BuildTools/Converters/Readers/OpenCLReader.cs
Comment (Perksey) [inline src/Core/Silk.NET.BuildTools/Converters/Readers/OpenCLReader.cs:1122]: These two methods are genius! Can you move them into a common place (such as `Utilities`) so we can use them elsewhere in the generator? Preferably documented, too.

File: src/OpenCL/Extensions/Silk.NET.OpenCL.Extensions.INTEL/Enums/AdvancedMotionEstimationClMotionDetectDesc.gen.cs
Comment (Perksey) [inline src/OpenCL/Extensions/Silk.NET.OpenCL.Extensions.INTEL/Enums/AdvancedMotionEstimationClMotionDetectDesc.gen.cs:12]: It looks like the type name is actually cl_motion_detect_desc_intel and this is just sort of "namespaced" with the extension name. did we want to account for this?

File: src/OpenCL/Extensions/Silk.NET.OpenCL.Extensions.INTEL/Enums/AdvancedMotionEstimationClMotionDetectDesc2.gen.cs
Comment (Perksey) [inline src/OpenCL/Extensions/Silk.NET.OpenCL.Extensions.INTEL/Enums/AdvancedMotionEstimationClMotionDetectDesc2.gen.cs:12]: Do we need to care about .2?

File: src/OpenCL/Extensions/Silk.NET.OpenCL.Extensions.KHR/Enums/CommandBufferProperties.gen.cs
Comment (Perksey) [inline src/OpenCL/Extensions/Silk.NET.OpenCL.Extensions.KHR/Enums/CommandBufferProperties.gen.cs:16]: ngl I'm gushing over these changes, this looks so amazing and clean!

File: src/Core/Silk.NET.BuildTools/Converters/Readers/OpenCLReader.cs
Comment (Alexx999) [reply src/Core/Silk.NET.BuildTools/Converters/Readers/OpenCLReader.cs:808]: Yeah, it's basically constants, plus unnamed, plus bug in XML with cl_device_info that makes it unusable

File: src/Core/Silk.NET.BuildTools/Converters/Readers/OpenCLReader.cs
Comment (Alexx999) [reply src/Core/Silk.NET.BuildTools/Converters/Readers/OpenCLReader.cs:861]: Well, yeah, I've been thinking about better ways but...
Probably adding it to JSON spec same as renaming pass could be an option.

File: src/Core/Silk.NET.BuildTools/Converters/Readers/OpenCLReader.cs
Comment (Alexx999) [reply src/Core/Silk.NET.BuildTools/Converters/Readers/OpenCLReader.cs]: Done

File: src/OpenCL/Extensions/Silk.NET.OpenCL.Extensions.INTEL/Enums/AdvancedMotionEstimationClMotionDetectDesc.gen.cs
Comment (Alexx999) [reply src/OpenCL/Extensions/Silk.NET.OpenCL.Extensions.INTEL/Enums/AdvancedMotionEstimationClMotionDetectDesc.gen.cs:12]: Well, it's one of the parts where the spec is inconsistent, especially so extensions coming from Intel, those are the most "special". Basically, this is used only by Intel - can be used but only in a form of special "if vendor == Intel" handling.

File: src/OpenCL/Extensions/Silk.NET.OpenCL.Extensions.INTEL/Enums/AdvancedMotionEstimationClMotionDetectDesc2.gen.cs
Comment (Alexx999) [reply src/OpenCL/Extensions/Silk.NET.OpenCL.Extensions.INTEL/Enums/AdvancedMotionEstimationClMotionDetectDesc2.gen.cs:12]: Will check the extension spec

File: src/Core/Silk.NET.BuildTools/Converters/Readers/OpenCLReader.cs
Comment (Alexx999) [reply src/Core/Silk.NET.BuildTools/Converters/Readers/OpenCLReader.cs]: I do agree that it's not perfect but it's actually the last resort thing for stuff that's otherwise dangling.
For what it's worth, all this works back to Nov 2020 spec file, and the situation seems to be (slowly) improving.

Comment (Perksey) [reply src/Core/Silk.NET.BuildTools/Converters/Readers/OpenCLReader.cs]: I still think we should add extra attributes to the upstream spec rather than try and dissect a field with no defined format. This is what we've done in the past with OpenGL (we added the `group` attribute).

File: src/Core/Silk.NET.BuildTools/Converters/Readers/OpenCLReader.cs
Comment (Perksey) [inline src/Core/Silk.NET.BuildTools/Converters/Readers/OpenCLReader.cs:774]: Apologies for coming back with comments i didn't pick up on before... don't bitfields also constitute the flags attribute?

File: src/Core/Silk.NET.BuildTools/Converters/Readers/OpenCLReader.cs
Comment (Perksey) [inline src/Core/Silk.NET.BuildTools/Converters/Readers/OpenCLReader.cs:864]:
```suggestion
                            else if (typeStr?.StartsWith("OpenCL ") && typeStr.Contains("deprecated"))
                            {
                                // Core spec deprecation notices
                                var typeNameStart = typeStr.IndexOf(' ', "OpenCL ".Length) + 1;
                                var typeNameEnd = typeStr.IndexOf(' ', typeNameStart + 1);
                                type = new RenamedEntry(type.Original, typeStr.Substring(typeNameStart, typeNameEnd - typeNameStart));
                            }
                            else if (typeStr?.StartsWith("cl_uint ")) type = new RenamedEntry(type.Original, typeStr.Substring("cl_uint ".Length)); // QCOM extensions
```

File: src/Core/Silk.NET.BuildTools/Converters/Readers/OpenCLReader.cs
Comment (Perksey) [inline src/Core/Silk.NET.BuildTools/Converters/Readers/OpenCLReader.cs:898]:
```suggestion
                .Where(x => x.Attribute("comment")?.Value.Contains("deprecated"))
```

File: src/Core/Silk.NET.BuildTools/Converters/Readers/OpenCLReader.cs
Comment (Perksey) [inline src/Core/Silk.NET.BuildTools/Converters/Readers/OpenCLReader.cs:1046]:
```suggestion

```

**Comment by Perksey** [comment]
Happy to merge after the above.

**Comment by Alexx999** [comment]
those 2 are breaking changes - sadly, C# doesn't handle "bool?" type that well

**Comment by Perksey** [comment]
Ah, in that case our preferred style is “?? false” instead of “== true” for a nullable LHS.

**Comment by Alexx999** [comment]
OK, done

**Review by Perksey — APPROVED** [review]
Thank

**Comment by Perksey** [comment]
Looks like CI is failing…

**Comment by Alexx999** [comment]
Simply re-running it fixed the issue, also, the previous build was OK (without the "== true" -> "?? false" change), so looks like a build system glitch to me

**Pull request merged**
